website/oauth/models/__init__.py

```diff
@@ -0,0 +1,325 @@
+    provider = fields.StringField(required=True)
+
+    # The unique, persistent ID on the remote service.
+    provider_id = fields.StringField()
+
+    display_name = fields.StringField()
+
+    def __repr__(self):
+        return '<ExternalAccount: {}/{}>'.format(self.provider,
+                                                 self.provider_id)
+
+
+class ExternalProviderMeta(abc.ABCMeta):
+
+    def __init__(cls, name, bases, dct):
+        super(ExternalProviderMeta, cls).__init__(name, bases, dct)
+        if not isinstance(cls.short_name, abc.abstractproperty):
+            PROVIDER_LOOKUP[cls.short_name] = cls
+
+
+class ExternalProvider(object):
+    """
+
+    """
+
+    __metaclass__ = ExternalProviderMeta
+
+    account = None
+    _oauth_version = OAUTH2
+
+    def __repr__(self):
+        return '<{name}: {status}>'.format(
+            name=self.__class__.__name__,
+            status=self.account.provider_id if self.account else 'anonymous'
+        )
+
+    @abc.abstractproperty
+    def auth_url_base(self):
+        """The base URL to begin the OAuth dance"""
+        raise NotImplementedError
+
+    @property
+    def auth_url(self):
+        """The URL to begin the OAuth dance.
+
+        Accessing this property will create an ``ExternalAccount`` if one is not
+        already attached to the instance, in order to store the state token and
+        scope."""
+
+        session = get_session()
+
+        # create a dict on the session object if it's not already there
+        if session.data.get("oauth_states") is None:
+            session.data['oauth_states'] = {}
+
+        if self._oauth_version == OAUTH2:
+            # build the URL
+            oauth = OAuth2Session(
+                self.client_id,
+                redirect_uri=web_url_for('oauth_callback',
+                                         service_name=self.short_name,
+                                         _absolute=True),
+                scope=self.default_scopes,
+            )
+
+            url, state = oauth.authorization_url(self.auth_url_base)
+
+            # save state token to the account instance to be available in callback
+            session.data['oauth_states'][self.short_name] = {'state': state}
+
+        elif self._oauth_version == OAUTH1:
+            # get a request token
+            oauth = OAuth1Session(
+                client_key=self.client_id,
+                client_secret=self.client_secret,
+            )
+
+            response = oauth.fetch_request_token(self.request_token_url)
+
+            session.data['oauth_states'][self.short_name] = {
+                'token': response.get('oauth_token'),
+                'secret': response.get('oauth_token_secret'),
+            }
+
+            url = oauth.authorization_url(self.auth_url_base)
+
+        return url
+
+    @abc.abstractproperty
+    def callback_url(self):
+        """The provider URL to exchange the code for a token"""
+        raise NotImplementedError()
```

Comment:
Very minor, but I'm used to seeing `abstractproperty` OR `raise NotImplementedError`, not both (with the body of the method being `pass` or `return`). Just curious--why use both?